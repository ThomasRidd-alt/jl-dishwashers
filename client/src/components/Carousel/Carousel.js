import React, {Component} from 'react';
import SquareImage from "../SquareImage";
import chevronLeft from '../../images/chevron-left.svg'
import chevronRight from '../../images/chevron-right.svg'

class Carousel extends Component {

    constructor(props) {
        super(props);
        this.images = props.images ? props.images : [];
        this.state = {index: 0};
    }

    prevSlide = () => {
        this.setState({index: this.state.index > 0 ? this.state.index - 1 : this.images.length - 1})
    }

    nextSlide = () => {
        this.setState({index: this.state.index < this.images.length - 1 ? this.state.index + 1 : 0})
    }

    render() {
        return (
            <div className={'jl-carousel'} aria-hidden={'true'}>
                <div className={'jl-carousel__image-container'}>
                    {this.images.map((image, index) => {
                        return (
                            <div
                                className={index === this.state.index ? 'jl-carousel__slide' : 'jl-carousel__slide jl-carousel__slide--hidden'}
                                key={`image-${index}`}>
                                <SquareImage src={image} alt={''}/>
                            </div>
                        )
                    })}
                    <button className={'jl-carousel__chevron-container jl-carousel__previous'} onClick={this.prevSlide}>
                        <img className={'chevron'} src={chevronLeft} alt={''}/>
                    </button>
                    <button className={'jl-carousel__chevron-container jl-carousel__next'} onClick={this.nextSlide}>
                        <img className={'chevron'} src={chevronRight} alt={''}
                        />
                    </button>
                </div>
                <div className={'jl-carousel__pip-container'}>
                    {this.images.map((image, index) => {
                        return (
                            <span
                                className={`jl-carousel__pip ${index === this.state.index ? 'jl-carousel__pip--solid' : ''}`}
                                key={`pip-${index}`}/>
                        )
                    })}
                </div>
            </div>
        )
    }
}

export default Carousel;