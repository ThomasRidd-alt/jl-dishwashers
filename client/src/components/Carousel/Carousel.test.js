import React from "react";
import {shallow} from 'enzyme';
import Carousel from "./Carousel";

describe('Carousel', () => {
    const id = 1;
    const images = ['image1', 'image2', 'image3']
    let component = shallow(<Carousel images={images}/>);

    beforeEach(() => {
        component = shallow(<Carousel images={images}/>);
    })

    it('renders properly', () => {
        expect(component).toMatchSnapshot()
    })

    it('renders the same number of frames as urls', () => {
        expect(component.find('.jl-carousel__slide').length).toEqual(images.length)
    })

    const visibleSlides = (component) => {
        return component.find('.jl-carousel__slide').length - component.find('.jl-carousel__slide--hidden').length
    }

    it('renders only one without the hidden class', () => {
        expect(visibleSlides(component)).toEqual(1)
    })

    it('starts with index 0', () => {
        expect(component.state().index).toEqual(0)
    })

    it('increments index on click next', () => {
        component.find('.jl-carousel__next').simulate('click');

        expect(component.state().index).toEqual(1)
    })

    it('decrements index on click previous', () => {
        component.find('.jl-carousel__next').simulate('click');
        expect(component.state().index).toEqual(1)

        component.find('.jl-carousel__previous').simulate('click');
        expect(component.state().index).toEqual(0);
    })


    it('loops back to end on click previous when at start', () => {
        component.find('.jl-carousel__previous').simulate('click');
        expect(component.state().index).toEqual(images.length - 1)
    })

    it('loops back to start when at the end', () => {
        component.find('.jl-carousel__previous').simulate('click');
        expect(component.state().index).toEqual(images.length - 1)

        component.find('.jl-carousel__next').simulate('click');
        expect(component.state().index).toEqual(0);
    })
})