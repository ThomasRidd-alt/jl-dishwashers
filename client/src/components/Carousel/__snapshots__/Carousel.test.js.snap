// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders properly 1`] = `ShallowWrapper {}`;
